File: demo/src/main/java/hu/telekom/ubuntureleasedemo/java/UbuntuReleaseController.java
package hu.telekom.ubuntureleasedemo.java;

import org.jetbrains.annotations.NotNull;
import org.reactivestreams.Publisher;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.Random;
import java.util.function.Function;

import static hu.telekom.ubuntureleasedemo.kotlin.WireMockConfigKt.RELEASE_COUNT;

@RestController
public class UbuntuReleaseController {

    private static final Logger LOG = LoggerFactory.getLogger(UbuntuReleaseController.class);

    private UbuntuReleaseClient ubuntuReleaseClient;

    public UbuntuReleaseController(UbuntuReleaseClient ubuntuReleaseClient) {
        this.ubuntuReleaseClient = ubuntuReleaseClient;
    }

    @GetMapping("/random-releases")
    public Flux<String> getReleaseNames(@RequestParam(name = "qty", defaultValue = "3") int qty) {
        return Flux.range(0, qty)
            .map(mapToRandomIndex())
            .flatMap(fetchVersion())
            .flatMap(ver -> fetchFirstCode(ver)
                .zipWith(fetchLastCode(ver)))
            .map(tuple -> tuple.getT1() + " " + tuple.getT2())
            .flatMap(fetchLts())
            .map(it -> it + "<br/>\r\n");
    }

    @NotNull
    private Function<Integer, Integer> mapToRandomIndex() {
        return count -> {
            int idx = new Random().nextInt(RELEASE_COUNT);
            LOG.info("Mapped count {} to index {}", count, idx);
            return idx;
        };
    }

    @NotNull
    private Function<Integer, Publisher<? extends String>> fetchVersion() {
        return idx -> ubuntuReleaseClient.getVersion(idx)
            .doOnNext(it -> LOG.info("> {}: received version {}", idx, it));
    }

    @NotNull
    private Mono<String> fetchFirstCode(String ver) {
        return ubuntuReleaseClient.getFirstCodeName(ver)
            .doOnNext(it -> LOG.info(">> {}: received first codename {}", ver, it));
    }

    @NotNull
    private Mono<String> fetchLastCode(String ver) {
        return ubuntuReleaseClient.getLastCodeName(ver)
            .doOnNext(it -> LOG.info(">>> {}: received last codename {}", ver, it));
    }

    @NotNull
    private Function<String, Mono<? extends String>> fetchLts() {
        return codename -> ubuntuReleaseClient.getLts(codename)
            .onErrorResume(it -> Mono.just(""))
            .doOnSuccess(it -> LOG.info(">>>> {} {}", codename, it))
            .map(it -> codename + " " + it);
    }

}

File: demo/src/main/kotlin/UbuntuReleaseDemo/UbuntuReleaseController.kt
package hu.telekom.ubuntureleasedemo.kotlin

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.newFixedThreadPoolContext
import org.slf4j.LoggerFactory
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.reactive.function.client.WebClientResponseException
import kotlin.random.Random

@RestController
class UbuntuReleaseController(val ubuntuReleaseClient: UbuntuReleaseClient) {

    val log = LoggerFactory.getLogger(UbuntuReleaseController::class.java);

    val ctx = newFixedThreadPoolContext(50, "custom-tp")

    @GetMapping("/random-releases")
    fun getReleaseNames(@RequestParam(name = "qty", defaultValue = "3") qty: Int): Flow<String> {
        return flow {
            (0 until qty)
                    .map { mapToRandomIndex(it) }
                    .map { fetchVersion(it) }
                    .map {
                        val ver = it.await()
                        fetchFirstCode(ver) to fetchLastCode(ver)
                    }
                    .map {
                        val codename = "${it.first.await()} ${it.second.await()}"
                        fetchLts(codename)
                    }
                    .map { emit(it.await() +  "<br/>\r\n") }
        }
    }


    private fun mapToRandomIndex(it: Int): Int {
        val idx = Random.nextInt(RELEASE_COUNT)
        log.info("Mapped count {} to index {}", it, idx)
        return idx
    }

    private suspend fun fetchVersion(idx: Int) = CoroutineScope(ctx).async {
        val ver = ubuntuReleaseClient.getVersion(idx)
        log.info("> {}: received version {}", idx, ver)
        ver
    }

    private suspend fun fetchFirstCode(ver: String) = CoroutineScope(ctx).async {
        val name = ubuntuReleaseClient.getFirstCodeName(ver)
        log.info(">> {}: received first codename {}", ver, name)
        name
    }

    private suspend fun fetchLastCode(ver: String) = CoroutineScope(ctx).async {
        val name = ubuntuReleaseClient.getLastCodeName(ver)
        log.info(">>> {}: received last codename {}", ver, name)
        name
    }

    private suspend fun fetchLts(codename: String) = CoroutineScope(ctx).async {
        val lts = try { ubuntuReleaseClient.getLts(codename) }
        catch (ex: WebClientResponseException.NotFound) { "" }

        log.info(">>>> {} {}", codename, lts)
        "$codename $lts"
    }

}
